enhancements/installer/component-selection.md

```diff
@@ -0,0 +1,423 @@
+
+```yaml
+InstallConfig.Capabilities
+  Default: [include|exclude]
+  Exclude:
+  - console
+  - samples
+  Include:
+  - registry
+```
+
+Which resources ultimately get installed for a given cluster would be the set of resources encompassed
+by the CLUSTER_PROFILE(if any), minus any resources explicitly excluded by the excluded capabilities configuration,
+plus any explicitly included capabilities.  Any resource that is not included by the CLUSTER_PROFILE, is not
+included via the Include stanza, and is not Excluded via the Exclude stanza, will be filtered or not based on
+the `Default` value (include or exclude).
+
+The `Default` field (carried through to the CVO) allows an Admin to choose what behavior they want for new
+capabilities that might come as part of a future upgrade (either have the new capability installed during
+the upgrade, or exclude it)
+
+Specifying the same capability as an `include` and an `exclude` is an error condition.
+
+
+Examples of candidate components to be treated as capabilities:
+
+* console
+* imageregistry
+* samples
+* cluster baremetal operator
+* olm/marketplace
+* csi-*
+* insights
+* monitoring
+* ???
+
+
+
+3) Pass the list of filtered annotations to the CVO.  This is distinct from overrides because overrides
+put the cluster in an unsupported state.  Filtered annotations are supported for upgrades.  The filtered
+components will be listed in the [ClusterVersion](https://github.com/openshift/api/blob/4436dc8be01e8dcd8b250e1b32bb0fbd64ba78ac/config/v1/types_cluster_version.go#L35) object:
+
+```yaml
+  spec:
+    capabilities:
+      default: [include|exclude]
+      exclude:
+      - console
+      - samples
+      include:
+      - registry
+```
+
+The CVO will filter out(not apply/reconcile) resources that are annotated with `capability.openshift.io/$exclusions`
+and listed in the `exclude` section, or are not listed and the `default` setting is `exclude`.  In addition
+any explicitly included capability should be added, regardless of the CLUSTER_PROFILE.
+
+If a resource has multiple `capability.openshift.io` annotations, then the resource will be filtered if any
+of the annotations on the resource are matched by a configured filter.  This allows a component to define
+a `ServiceMonitor` which is part of both the `Monitoring` capability and the `Foo` capability.  If either
+of those capabilities are excluded, the `ServiceMonitor` should not be created.  (If Monitoring is excluded
+then no ServiceMonitors should be created, if only Foo is excluded, then only Foo's ServiceMonitors should
+be excluded).
+
+Specifying the same capability as an `include` and an `exclude` is an error condition.
+
+Note: The CVO must wait until the ClusterVersion resource exists (created by the installer) to ensure
+there is no race condition in which the CVO starts creating resources before it has the filter list.
```

Comment:
nit: the CVO only needs to defer capability-linked resources.  There will be some resources associated with the configured cluster profile which it will know it can apply, regardless of the configured capabilities.

Reply:
unless we think there will be a need to apply those (for bootstrapping purposes), it seems simpler to just not apply anything until the CV exists, right?

Reply:
Simpler yes.  But [possibly significantly slower][1] (or maybe not, I've only checked timings in those two jobs).  As far as the enhancement proposal goes, this is an internal CVO detail, so perhaps reword to give the CVO room to wiggle between possible implementations?  Or we can experiment in the CVO and then bump this enhancement doc later if we need to catch it back up with whatever the eventual reality ends up being.

[1]: https://github.com/openshift/cluster-version-operator/pull/682#issuecomment-958656065

Reply:
done (new commit)